### DISCUSSION TITLE
OPCUA real-time data

### DISCUSSION THREAD

**Discussion by PetterBMarkussen** [admin-shell-io/questions-and-answers#75] [Q&A]
Hi, 

I have been using the AASX Package explorer to generate a serialized OPCUA nodeset2.xml file. I have some question regarding the serialization 

It does not seam that the OPC UA companion specification for AAS utilizes the native OPC UA variable concept. How should real-time data in an OPC UA server be used in the context of AAS? I see that the serialization from AAS -> nodeset2.xml only uses OPC UA properties. Real time data on an OPC UA server should be implemented by using OPC UA native "variable" concept. Could someone please clearify why this concept has not been used in the serlization? For analogue measurment values, the variables concept is used in OPC UA. AnalogueItemType is the usuall variable type to use in OPC UA for exposing a analogue signal on the server. This variable also includes several standard properties s.a. EURange and EUInformation. I dont see any mention of utilizing these key standard OPC UA concepts in this searilization.

**Comment by GoetzGoerisch** [comment]
For realtime information the applicable OPC UA companion specification should be used.

**Reply by PetterBMarkussen** [reply]
Ok, thanks for fast answer.  Do you know where this is specified in the AAS standard or documents?

**Reply by GoetzGoerisch** [reply]
As far as I'm aware this is content of the Submodel Template OPC UA Server Datasheet.

**Reply by BFrKUKA** [reply]
For the topic "AASX Package explorer to generate a serialized OPCUA nodeset2.xml file" please contact the AASX Package explorer developer. Currently I lack in understanding why to "serialize an AAS to an OPC UA nodeset file"?

I am the editor of the AAS submodel template "OPC UA Server Datasheet" and we want to provide with this submodel template the possibility to provide information about an OPC UA Server integrated in "any product" (e.g. a machine / a device)
There we want to cover the descriptions of namespaces (for types and instances.

Regarding "real-time-data" my question is the requirement on "real-time".
OPC UA Servers provide connectivity for clients, which i would say -> this is no realtime. (you geht data incl., timestamps... but not "in real-time")
OPC UA Servers may support OPC UA Pub/Sub (with MQTT(Json) or in future TSN)... this is just for me "real-time-data"... to cover all this aspects of "real-time-data" in OPC UA, there is the OPC UA FLC initiative which defines UAFX as a "framework" to establish OPC UA Pub/Sub Connections... but not for AAS... but OPC UA Servers ...

**Reply by PetterBMarkussen** [reply]
Thanks for information.

 Can you confirm that AAS is only intended to be used with static values and not live values?

**Reply by StenGruener** [reply]
@BFrKUKA "Exporting to nodeset2" means generating of AAS representation according to "OPC 30270 – OPC UA for I4 Asset Administration Shell" companion spec, this is not to be mixed with AAS submodel template "OPC UA Server Datasheet" (which AFAIK represents a runtime interface of an OPC UA server). Again, using OPC 30270 you can represent any AAS as a native OPC UA model and then (in theory) use PubSub etc.

@PetterBMarkussen the aim of the "OPC 30270 – OPC UA for I4 Asset Administration Shell" is to map the AAS meta-model to the OPC UA "world". 
My personal opinion as early participant in some of OPC 30270 meetings: Since AAS and OPC UA modeling best practices overlap partially. For example, AAS is currently missing *exact* concepts of EURange and EUInformation (e.g. it has units, but the ranges are AFAIK not specified clearly and can be modeled in some "gray-zones" like qualifiers). Same holds for the mainstream interfaces used in UA specs like DI-based IVendorNameplate etc. and the identification submodel of the AAS. Therefore, a best-achievable compromise between partially overlapping AAS concepts and concepts of OPC UA was reached (remember, you also need to be able to map AAS not only to OPC UA models, but to JSON, RDF, etc.).
Most probably, this question can be pointed to spokesman of the OPCF working group (https://opcfoundation.org/markets-collaboration/i4aas/).

@PetterBMarkussen AAS is not limited to static values at all ;)

Is it helpful?

**Reply by BFrKUKA** [reply]
@StenGruener - and what to do, with the exported NodeSet? ... Typically a NodeSet in OPC UA means is used, when I build up an OPC UA Server with an OPC UA SDK... Loading the NodeSet2.XML (e.g. from OPC 30270) and get classes "for my application which runs the OPC UA Server" to instanciate the types defined by the NodeSet2. 
 (and to to say "in theory with PubSub" is "a good joke", because OPC UA PubSub is just possible to provide by each OPC UA Server, which provides a Publisher, .. this has nothing to do with OPC 30270"

@PetterBMarkussen I was fighting long time to explain the people, that "modelling with OPC UA" is just another story then "modelling AAS and submodels"... and that a "merge/remodelling" is meaningless in my view... and not possible or just very very complex, because of different concepts (OPC UA AnalogItemType with EU, Range, etc... and AAS-Submodel Element Property (Range) witch Qualifiers (AllowedRange) or ConceptDescription to provide an Engineering Unit... 

I do not disagree, that in an AAS are also "live (changing) values" - but in my view not "each runtime data of a real device" (in an OPC UA Server our product has about 7000 nodes!) is meaningfull to "updtae in an AAS" - who will pay the bill for data exchange with the cloud provider?

Back to the SMT AAS OPC UA Server Datasheet: we want to provide all information about an OPC UA Server delivered (as or) in  a product. This can be used by "any application", which has access to the AAS and this submodel. 
We are working on concepts how to "establish" connectivity of an AAS to "its asset with an OPC UA Server interface" to **enhance** AAS-submodles with runtime dat which is needed for this submodel. (e.g. EnergyConsumption of the Asset to calculate PCF values)
In a nutshell the AAS SMT OPC UA Server Datasheet should have all information which is needed to establish a client to connect to an server and which information and services there (the client) might use.... where this client passes the information to is "irrelevant"... it can be passed to any application or to any submodel of an AAS. (in my view)

**Reply by StenGruener** [reply]
@BFrKUKA NodeSets can contain both types and instances since both are nodes. So basically, yes, import such a mixed Nodeset or multiple Nodeset files into an OPC UA server and "wire up" instance variables with some backend. This is at least how the mentioned AASXPackageExplorer demo is working - it exports both types from the companion spec and instances with are specific to the currently viewed AAS.

Personally, I fully support the submodel definition of OPC UA endpoints as it follows the same concepts we have proposed for AML engineering data in our MTP model template (so, do not remodel, but embed and use aas for cross-references). Also, I personally agree that blind re-modeling of everything in the domain of AAS is probably more harmful than useful.

Still, I am also aware of some use-cases to model (and esp. retrofit) devices as AAS for the sake of state-based-control (think of PackML) which have no real hard real-time requirement. Reference here is the BaSys 4.2 project.

Also, the OPC UA companion spec exists as a matter of fact, and can in theory be used for some use-cases (I am currently not aware of any, though). I will probably put this question in our expert group, maybe some one else is aware of some usage of the companion spec in the wild. Are you aware of any?

**Reply by BFrKUKA** [reply]
@StenGruener 
1. aggree... Nodeset2 contain instances as well... but I lack in undertanding in "how the AASX PackageExplorer demo" is working for" wiring up instance variables with some backend"

2. Regarding MTP AAS -> I have seen ONE OPC UA endpoint descritpion (als EndpointUri) in this submodel template. in my view this information is (maybe) not enough, that it can be used meaningfull (without information about security credendials, which must apply to this EndointURL or the different protocolls and encoding variants (UATCP, HTTPS, Websockets, ..)
Regarding AML I am just a noob and maybe want to understand this better

3, Retrofit: in my view with OPC UA there are enough soultions (OPC Router, ...)  on the market for "retrofitting" existing devices which provide "brownfield" communcation interfaces (ProfiNet, IO-Link, Hart, ...) and for a lot of them OPC UA CS exist, to describe the "retrofitted technolgoy"

4: PackMl as a OPC UA CS defines, how this should be implemented in a OPC UA Server of a packaging machine... in a brownfield system maybe on a "gateway solution".
In the OPC UA CS PackMal is described how to control via OPC UA mechanisms (e.g. controlling a statemachine with triggering transitions via methods)... 
 So why to "remodel PackML" in an AAS Submodel? - with which technology to connect to the real machine? "Industry 4.0 language" to "control the behavior of the real machine"... at the end the machine may speak OPC UA, Profinet, EtnernetIP, CCLink, Ehtercat or Modbus TCP... 

5. Existing OPC UA CS: 
They are not only a fact: they are defined by technology specific specialists of different verndors or users of products of this technology in industry -> and they integrate the functions, they define within their CS in the products -> OPC UA servers! 

I have seen several implementation attempts where an OPC UA (Companion) Specification (DI, Robotics) was tried to be imported into the AASX Explorer in a submodel. I (as the editor of OPC UA Robotics and the chair of this group) did not understand, "what AAS remodelling was doing with the "OPC UA model".... in our view the result (of the AASX import) is not useable at all.

Regarding use cases with OPC UA CS:
OPC UA Machinery CS provides a description how "Nameplate information" of an asset is findable inside an OPC UA Server (and which Properties are provided). In my view when I want to "onboard or register" an real device to a cloud (and an AAS) I want to look up, if the real asset fits to the digital twin (AAS)... so I typically compare the "AAS SM Nameplate data" with "OPC UA Machinery Identification data" I find in the OPC UA server.

**Reply by PetterBMarkussen** [reply]
> @StenGruener - and what to do, with the exported NodeSet? ... Typically a NodeSet in OPC UA means is used, when I build up an OPC UA Server with an OPC UA SDK... Loading the NodeSet2.XML (e.g. from OPC 30270) and get classes "for my application which runs the OPC UA Server" to instanciate the types defined by the NodeSet2. (and to to say "in theory with PubSub" is "a good joke", because OPC UA PubSub is just possible to provide by each OPC UA Server, which provides a Publisher, .. this has nothing to do with OPC 30270"
> 
> @PetterBMarkussen I was fighting long time to explain the people, that "modelling with OPC UA" is just another story then "modelling AAS and submodels"... and that a "merge/remodelling" is meaningless in my view... and not possible or just very very complex, because of different concepts (OPC UA AnalogItemType with EU, Range, etc... and AAS-Submodel Element Property (Range) witch Qualifiers (AllowedRange) or ConceptDescription to provide an Engineering Unit...
> 
> I do not disagree, that in an AAS are also "live (changing) values" - but in my view not "each runtime data of a real device" (in an OPC UA Server our product has about 7000 nodes!) is meaningfull to "updtae in an AAS" - who will pay the bill for data exchange with the cloud provider?
> 
> Back to the SMT AAS OPC UA Server Datasheet: we want to provide all information about an OPC UA Server delivered (as or) in a product. This can be used by "any application", which has access to the AAS and this submodel. We are working on concepts how to "establish" connectivity of an AAS to "its asset with an OPC UA Server interface" to **enhance** AAS-submodles with runtime dat which is needed for this submodel. (e.g. EnergyConsumption of the Asset to calculate PCF values) In a nutshell the AAS SMT OPC UA Server Datasheet should have all information which is needed to establish a client to connect to an server and which information and services there (the client) might use.... where this client passes the information to is "irrelevant"... it can be passed to any application or to any submodel of an AAS. (in my view)

Ok, thanks for this information. The reason I ask about this OPC UA vs AAS, is that I came across a paiper that has used the OPC 30270 standard to parse a AAS file -> nodeset2 file. This author had then used the properties in the OPC UA information model, to subscribe to live values. This is where I am sceptic.  The definition of an property in OPC UA context is something that describes a characteristic of a node in the adress space. It should not be used to subscribe to live values, even though the author did exactly this.

**Reply by StenGruener** [reply]
```
Regarding use cases with OPC UA CS:
OPC UA Machinery CS provides a description how "Nameplate information" of an asset is findable inside an OPC UA Server (and which Properties are provided). In my view when I want to "onboard or register" an real device to a cloud (and an AAS) I want to look up, if the real asset fits to the digital twin (AAS)... so I typically compare the "AAS SM Nameplate data" with "OPC UA Machinery Identification data" I find in the OPC UA server.
```
@BFrKUKA In our domain, PA-DIM, we typically assume that ProductInstanceUri from part -100 in the OPC UA device is exactly the "AssetID" in the AAS-Domain and also what is printed on the device (identification link). I guess also that this is a common understanding wrt. your use-case.

**Reply by BFrKUKA** [reply]
@StenGruener the VDMA Machinery CS also bases for this on the IVendorNameplateType of part 100, which provides the ProductInstanceUri.  In the same way PA-DIM does.
And currently I just have this question, "where in the AAS" is the correllated "ProductInstanceUri" defined?
I currently "try to model with AASX Explorer" and as understood I define on the Asset the "Identifiable id as an IRI" and the AAS points via the Asset Reference to the Asset. - Am I right?

**Reply by StenGruener** [reply]
@BFrKUKA correct. You can also resolve a known AssetID to multiple AAS using AAS registries. Maybe we can combine it with LDS-ME aswell? Maybe you can just give me a ping on firstname.lastname@de.abb.com?

**Reply by GEORGEANDREW5962** [reply]
Hi,
@PetterBMarkussen @GoetzGoerisch @StenGruener 

I am a beginer from Indonesia.
I already collect real time data from some machine PLC'S in DXP Takebishi OPC-UA Server and want to transfer its to the next layer.

Could you give any references how to transfer the real time data to AAS?
For example if I want to move the "actual current value" on DXP Server to the Operational data Submodels, "actual current value" element in AAS? 

It will be helpfull for me.

Thank you BR

**Reply by BFrKUKA** [reply]
Hi, at the IDTA currently we are working on a submodel template which might be usefull for your idea.
It is (currently) called asset interface description and currently has to parts:
1. to describe the communication technology of the "asset" (= the PLC with OPC UA Server) incl. connection information, but it should be able for other communication technologies as MQTT, ModbusTCP, ... as well)
2. to describe in detail the "real time data" (e.g. for OPC UA the NamespaceUri and the NodeId) and in which AAS-Submodel-Element it should be mapped.
So this is just regarding how we want to standardize such AAS Submodels at the IDTA. But there we have just discussions because of overlaps to the SubmodelTemplate "OPC UA Server DataSheet", which would have all "connection information" for an OPC UA Server...

But for a practical approach I think the way Basys project did it (as I know with OPC UA and MQTT) maybe @StenGruener can help you.

Regards, Bernd
